mitodl/ocw-hugo-themes
# offline video playback

#### Pre-Flight checklist


#### What are the relevant tickets?
Closes https://github.com/mitodl/ocw-hugo-themes/issues/1133

#### What's this PR do?
This PR changes the behavior video pages and embeds in the offline themes. A new partial, `video_player.html` was added. Places that called `youtube_player.html` directly have been altered to call this partial instead. In the online theme, `youtube_player.html` is simply called without any other action. In the offline theme, a check is made against the `static_resources` leaf node to see if a copy of the archive video is present at build time. If the video exists, `local_video_player.html` is called, which renders a VideoJS player pointing at the locally stored archive video. If the video is not present, `youtube_player.html` is called instead to load the video from YouTube.

#### How should this be manually tested?
 - Download the course content for [18-06-fall-2010](https://github.mit.edu/ocw-content-rc/18.06-spring-2010)
 - Make sure you have [`ocw-hugo-projects`](https://github.com/mitodl/ocw-hugo-projects) cloned locally
 - Source `ocw-studio` production AWS credentials and set them up in a terminal
 - In the same terminal change directory to the course content repo you downloaded above
 - Run:
```
mkdir -p content/static_resources
aws s3 sync s3://ol-ocw-studio-app-production/courses/18-06-linear-algebra-spring-2010/ ./content/static_resources/
```
 - Once the static resources have been downloaded, run the following in the `ocw-hugo-themes` directory, being sure to use *absolute paths*: `yarn build /path/to/18.06-spring-2010/ /path/to/ocw-hugo-projects/ocw-course-v2/config-offline.yaml`
 - In your file browser, go to your course content repo and enter the `dist` folder, then double click on `index.html` to open the offline site in a web browser
 - Go to the Instructor Insights page and click on the embedded video there and ensure it plays back and that the VideoJS controls work properly and that the video is sized properly in the layout
 - Go to the Video Lectures section and ensure the same thing for any of the video lectures contained within
 - Go back to the course content repo, enter the `content/static_resources` folder and delete one of the archive videos, noting which one you deleted. Find the page that this video is referenced on so you know the URL to browse to when testing by searching the repo for the video filename.
 - Re-run the offline build
 - Visit the page for the video that is now missing and verify that the YouTube player is loaded instead.

#### Any background context you want to provide?
This is part of the [mirror drive epic](https://github.com/mitodl/hq/issues/785) and enables playback of archive videos if they are included with a course at build time. This PR should be reviewed, but not merged until a separate situation is addressed. In `ocw-studio` where MIT's OCW content is authored and subsequently Concourse where the builds are actually run, the offline build is run and is both used to generate the individual ZIP downloads of courses as well as the content for the mirror drive, stored in separate S3 buckets. In our Concourse pipelines, mp4 files are currently filtered out of the ZIP downloads to reduce the file size of the downloads. Before this is merged we should alter the logic here to run a Hugo build with *and* without the archive videos. The version without the archive videos should have the videos filtered out before the Hugo build even runs, so that the theme does not assume they exist and renders the YouTube player instead on pages in that offline site that have video players. If this PR is merged without addressing this issue first, the offline sites in ZIP downloads will have offline video players rendered that will attempt to load local mp4 videos that don't exist.

This should not be merged until https://github.com/mitodl/ocw-studio/pull/1836 is released, as the way mp4 files are filtered out of builds in Concourse after Hugo runs means that local video players will be rendered in offline sites, and the videos will not be present in the ZIP file.

#### Screenshots (if appropriate)
![image](https://github.com/mitodl/ocw-hugo-themes/assets/12089658/5760a7e8-feae-415a-95fc-b7dd62ae96ec)
